lib/vrl/compiler/src/enrichment_tables.rs

```diff
@@ -0,0 +1,44 @@
+use dyn_clone::DynClone;
+
+pub enum Condition {
+    Equals { field: String, value: String },
+}
+
+pub trait EnrichmentTableSetup {
+    fn table_ids(&self) -> Vec<String>;
+    fn add_index(&mut self, table: &str, fields: &[&str]) -> Result<(), String>;
+}
+
+pub trait EnrichmentTableSearch: DynClone {
+    fn find_table_row(
+        &self,
+        table: &str,
+        criteria: Vec<Condition>,
+    ) -> Result<Option<Vec<String>>, String>;
+}
+
+dyn_clone::clone_trait_object!(EnrichmentTableSearch);
```

Comment:
I never really understood why we need this. We add this for all components as well (and any downstream types). Do you have any idea? Just curious if there is still a good reason to keep this around, or if we could file an issue to work on removing this requirement.

Reply:
The only reason this is needed is because `state::Compiler::snapshot` needs it to clone a snapshot.

The underlying implementation in Vector makes these clones very cheap, although that would not necessarily be a guarantee for other consumers of VRL..

Reply:
Oh no, sorry, that's why `EnrichmentTableSetup` needs Clone. Search needs clone because it is a member of `Program` which wants Clone.

I can't think of a reason why we couldn't wrap that in an `Arc` which could remove that requirement.. 

It would mean the concrete implementation would be something like `Arc<Arc<ArcWrap<Arc<EnrichmentSearch>>>>`...

Reply:
My primary question is around the use of `dyn_clone`. I think it got introduced somewhere around when we worked on the Lookup stuff? But I might be misremembering.

Reply:
Ah, it's because the tables are being stored as a `Box<dyn TableSearch>`.  We could likely change that and make it `Compiler<T> where T: TableSearch`. That's worth looking into.